storage/innobase/row/row0ins.cc

```diff
@@ -1755,11 +1755,90 @@ row_ins_check_foreign_constraint(
 		cmp = cmp_dtuple_rec(entry, rec, offsets);
 
 		if (cmp == 0) {
-			if (rec_get_deleted_flag(rec,
-						 rec_offs_comp(offsets))) {
+			if (DICT_TF2_FLAG_IS_SET(check_table, DICT_TF2_VERSIONED)) {
+				trx_id_t end_trx_id = 0;
+				ulint len;
+
+				if (dict_index_is_clust(check_index)) {
+					ulint nfield = dict_col_get_clust_pos(
+						&check_table->cols[check_table->vers_row_end],
+						check_index);
+
+					const byte *field = rec_get_nth_field(
+						rec,
+						offsets,
+						nfield, &len);
+					ut_a(len == 8);
+
+					end_trx_id = mach_read_from_8(field);
+				} else {
```

Comment:
Please, move body of this `else` to separate inline function (just before this function).